Mathlib/Tactic/DefeqTransformations.lean

```diff
@@ -0,0 +1,289 @@
+/-
+Copyright (c) 2023 Kyle Miller. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kyle Miller
+-/
+import Mathlib.Tactic.Basic
+
+/-! # Tactics that transform types into definitionally equal types
+
+This module defines a standard wrapper that can be used to create tactics that
+change hypotheses and the goal to things that are definitionally equal.
+
+It then provides a number of tactics that transform local hypotheses and/or the target.
+-/
+
+namespace Mathlib.Tactic
+
+open Lean Meta Elab Elab.Tactic
+
+/-- For the main goal, use `m` to transform the types of locations specified by `loc?`.
+If `loc?` is none, then transforms the type of target. `m` is provided with an expression
+with instantiated metavariables.
+
+`m` *must* transform expressions to defeq expressions. `runDefeqTactic` will throw
+an error if the result is not defeq. -/
+def runDefeqTactic (m : Expr → MetaM Expr)
+    (loc? : Option (TSyntax ``Parser.Tactic.location))
+    (tacticName : String) :
+    TacticM Unit := withMainContext do
+  withLocation (expandOptLocation (Lean.mkOptionalNode loc?))
+    (atLocal := fun h => liftMetaTactic1 fun mvarId => do
+      let ty ← instantiateMVars (← h.getType)
+      mvarId.changeLocalDecl' h (← m ty))
+    (atTarget := liftMetaTactic1 fun mvarId => do
+      let ty ← instantiateMVars (← mvarId.getType)
+      mvarId.change (← m ty))
+    (failed := fun _ => throwError "{tacticName} failed")
```

Comment:
```suggestion
`m` *must* transform expressions to defeq expressions. If `checkDefEq` is `true` (the default), 
`runDefeqTactic` will throw an error if the result is not defeq to the original expression. If 
`checkDefEq` is `false`, the result is still assumed to be defeq to the original, but this is not 
checked. -/
def runDefeqTactic (m : Expr → MetaM Expr)
    (loc? : Option (TSyntax ``Parser.Tactic.location))
    (tacticName : String)
    (checkDefEq := true) :
    TacticM Unit := withMainContext do
  withLocation (expandOptLocation (Lean.mkOptionalNode loc?))
    (atLocal := fun h => liftMetaTactic1 fun mvarId => do
      let ty ← instantiateMVars (← h.getType)
      mvarId.changeLocalDecl' h (← m ty) checkDefEq)
    (atTarget := liftMetaTactic1 fun mvarId => do
      let ty ← instantiateMVars (← mvarId.getType)
      mvarId.change (← m ty) checkDefEq)
    (failed := fun _ => throwError "{tacticName} failed")
```
Since this essentially maps `change`/`changeLocalDecl'` over the specified locations, and since both take the optional argument `checkDefEq := true`, I think it should probably preserve that option. Is there a reason we don't? Also, since `isDefEq` can be so expensive, it's probably useful to be able to turn checking off when we know the result is defeq (e.g., if `m` already checks for defeq); I think cases like this are quite frequent!

Maybe defeq checking should also be turned off for the specific tactics in this file that necessarily preserve defeq, like `whnf` and beta reduction.

Reply:
It seems good to do defeq checks if only because defeq is not a transitive relation, but I added the option. (I've imagined that it's probably not too expensive doing defeq checks for the sorts of transformations in this file, though I haven't measured anything. These are also probably going to be infrequently used tactics, so likely these defeq checks won't be measurable in a mathlib build.)